view/src/sync.rs

```diff
@@ -40,45 +44,67 @@ pub async fn scan_block(
-    let trial_decrypt = |note_payload: EncryptedNote| -> tokio::task::JoinHandle<Option<Note>> {
-        // TODO: change fvk to Arc<FVK> in Worker and pass to scan_block as Arc
-        // need this so the task is 'static and not dependent on key lifetime
-        let fvk2 = fvk.clone();
-        tokio::spawn(async move { note_payload.trial_decrypt(&fvk2) })
-    };
-
-    // Notes we've found in this block that are meant for us
-    let new_notes: Vec<SpendableNoteRecord>;
+    let trial_decrypt_note =
+        |note_payload: EncryptedNote| -> tokio::task::JoinHandle<Option<Note>> {
+            // TODO: change fvk to Arc<FVK> in Worker and pass to scan_block as Arc
+            // need this so the task is 'static and not dependent on key lifetime
+            let fvk2 = fvk.clone();
+            tokio::spawn(async move { note_payload.trial_decrypt(&fvk2) })
+        };
+    // Trial-decrypt a swap with our own specific viewing key
+    let trial_decrypt_swap =
+        |swap_payload: SwapPayload| -> tokio::task::JoinHandle<Option<SwapPlaintext>> {
+            // TODO: change fvk to Arc<FVK> in Worker and pass to scan_block as Arc
+            // need this so the task is 'static and not dependent on key lifetime
+            let fvk2 = fvk.clone();
+            tokio::spawn(async move { swap_payload.trial_decrypt(&fvk2) })
+        };
 
     // Nullifiers we've found in this block
     let spent_nullifiers: Vec<Nullifier> = nullifiers;
 
     // Trial-decrypt the notes in this block, keeping track of the ones that were meant for us
-    let decryptions = state_payloads
-        .iter()
-        .filter_map(|x| match x {
-            StatePayload::Note { note, source } => Some((trial_decrypt(note.clone()), source)),
-            _ => None,
-        })
-        .collect::<Vec<_>>();
-
-    let mut decrypted_applied_notes = BTreeMap::new();
-
-    for decryption in decryptions {
-        if let Some(note) = decryption.0.await.unwrap() {
-            decrypted_applied_notes.insert(note.commit(), note);
+    let mut note_decryptions = Vec::new();
+    let mut swap_decryptions = Vec::new();
+    let mut unknown_commitments = Vec::new();
+
+    for payload in state_payloads.iter() {
+        match payload {
+            StatePayload::Note { note, .. } => {
+                note_decryptions.push(trial_decrypt_note(note.clone()));
+            }
+            StatePayload::Swap { swap, .. } => {
+                swap_decryptions.push(trial_decrypt_swap(swap.clone()));
+            }
+            StatePayload::RolledUp(commitment) => unknown_commitments.push(commitment.clone()),
+        }
+    }
+    // Having started trial decryption in the background, ask the Storage for scanning advice:
+    let mut note_advice = storage.scan_advice(unknown_commitments).await?;
+    for decryption in note_decryptions {
+        if let Some(note) = decryption.await.unwrap() {
+            note_advice.insert(note.commit(), note);
+        }
+    }
```

Comment:
This is where the word "advice" comes from: the storage gives "scan advice" in the form of "known notes" that are then augmented with the results of trial decryption.